pkg/controller/daemon/daemon_controller.go

```diff
@@ -1000,108 +999,97 @@ func (dsc *DaemonSetsController) syncNodes(ds *apps.DaemonSet, podsToDelete, nod
-	for pos := 0; createDiff > pos; batchSize, pos = integer.IntMin(2*batchSize, createDiff-(pos+batchSize)), pos+batchSize {
-		errorCount := len(errCh)
-		createWait.Add(batchSize)
-		for i := pos; i < pos+batchSize; i++ {
-			go func(ix int) {
-				defer createWait.Done()
-				var err error
-
-				podTemplate := template.DeepCopy()
-				if utilfeature.DefaultFeatureGate.Enabled(features.ScheduleDaemonSetPods) {
-					// The pod's NodeAffinity will be updated to make sure the Pod is bound
-					// to the target node by default scheduler. It is safe to do so because there
-					// should be no conflicting node affinity with the target node.
-					podTemplate.Spec.Affinity = util.ReplaceDaemonSetPodNodeNameNodeAffinity(
-						podTemplate.Spec.Affinity, nodesNeedingDaemonPods[ix])
-
-					err = dsc.podControl.CreatePodsWithControllerRef(ds.Namespace, podTemplate,
-						ds, metav1.NewControllerRef(ds, controllerKind))
-				} else {
-					// If pod is scheduled by DaemonSetController, set its '.spec.scheduleName'.
-					podTemplate.Spec.SchedulerName = "kubernetes.io/daemonset-controller"
-
-					err = dsc.podControl.CreatePodsOnNode(nodesNeedingDaemonPods[ix], ds.Namespace, podTemplate,
-						ds, metav1.NewControllerRef(ds, controllerKind))
-				}
 
-				if err != nil && errors.IsTimeout(err) {
-					// Pod is created but its initialization has timed out.
-					// If the initialization is successful eventually, the
-					// controller will observe the creation via the informer.
-					// If the initialization fails, or if the pod keeps
-					// uninitialized for a long time, the informer will not
-					// receive any update, and the controller will create a new
-					// pod when the expectation expires.
-					return
-				}
-				if err != nil {
-					klog.V(2).Infof("Failed creation, decrementing expectations for set %q/%q", ds.Namespace, ds.Name)
-					dsc.expectations.CreationObserved(dsKey)
-					errCh <- err
-					utilruntime.HandleError(err)
-				}
-			}(i)
+	var (
+		errs         []error
+		createsCount int
+	)
+
+	createBarrier := controller.NewTaskBarrier()
+	// Batch the pod creates and fail on first error.
+	createBarrier.Limiter = controller.NewSlowStarter()
+	createBarrier.Policy = controller.ErrorPolicyFunc(func(err error) bool {
+		// If we see a timeout, it's possible that the pod is created but its
+		// initialization has timed out. If the initialization is successful
+		// eventually, the controller will observe the creation via the informer.
+		// If the initialization fails, or if the pod keeps uninitialized for a
+		// long time, the informer will not receive any update, and the controller
+		// will create a new pod when the expectation expires.
+		//
+		// Hence, we ignore timeouts.
+		if err != nil && !errors.IsTimeout(err) {
+			klog.V(2).Infof("Failed creation, decrementing expectations for set %q/%q", ds.Namespace, ds.Name)
+			dsc.expectations.CreationObserved(dsKey)
+			errs = append(errs, err)
```

Comment:
utilruntime.HandleError(err) is no longer needed ?

Reply:
This is double logging the error, but I'll leave it as is for now.